### PR TITLE
chore: prepare tokio-uring v0.4.0

### PR DESCRIPTION
**Pull request by FrankReh**
# 0.4.0 (November 5th, 2022)

### Fixed

- Fix panic in Deref/DerefMut for Slice extending into uninitialized part of the buffer ([#52])
- docs: all-features = true ([#84])
- fix fs unit tests to avoid parallelism ([#121])
- Box the socket address to allow moving the Connect future ([#126])
- rt: Fix data race ([#146])

### Added

- Implement fs::File::readv_at()/writev_at() ([#87])
- fs: implement FromRawFd for File ([#89])
- Implement `AsRawFd` for `TcpStream` ([#94])
- net: add TcpListener.local_addr method ([#107])
- net: add TcpStream.write_all ([#111])
- driver: add Builder API as an option to start ([#113])
- Socket and TcpStream shutdown ([#124])
- fs: implement fs::File::from_std ([#131])
- net: implement FromRawFd for TcpStream ([#132])
- fs: implement OpenOptionsExt for OpenOptions ([#133])
- Add NoOp support ([#134])
- Add writev to TcpStream ([#136])
- sync TcpStream, UnixStream and UdpSocket functionality ([#141])
- Add benchmarks for no-op submission ([#144])
- Expose runtime structure ([#148])

### Changed

- driver: batch submit requests and add benchmark ([#78])
- Depend on io-uring version ^0.5.8 ([#153])

### Internal Improvements

- chore: fix clippy lints ([#99])
- io: refactor post-op logic in ops into Completable ([#116])
- Support multi completion events: v2 ([#130])
- simplify driver operation futures ([#139])
- rt: refactor runtime to avoid Rc\<RefCell\<...>> ([#142])
- Remove unused dev-dependencies ([#143])
- chore: types and fields explicitly named ([#149])
- Ignore errors from uring while cleaning up ([#154])
- rt: drop runtime before driver during shutdown ([#155])
- rt: refactor drop logic ([#157])
- rt: fix error when calling block_on twice ([#162])

### CI changes

- chore: update actions/checkout action to v3 ([#90])
- chore: add all-systems-go ci check ([#98])
- chore: add clippy to ci ([#100])
- ci: run cargo test --doc ([#135])

[#52]: https://github.com/tokio-rs/tokio-uring/pull/52
[#78]: https://github.com/tokio-rs/tokio-uring/pull/78
[#84]: https://github.com/tokio-rs/tokio-uring/pull/84
[#87]: https://github.com/tokio-rs/tokio-uring/pull/87
[#89]: https://github.com/tokio-rs/tokio-uring/pull/89
[#90]: https://github.com/tokio-rs/tokio-uring/pull/90
[#94]: https://github.com/tokio-rs/tokio-uring/pull/94
[#98]: https://github.com/tokio-rs/tokio-uring/pull/98
[#99]: https://github.com/tokio-rs/tokio-uring/pull/99
[#100]: https://github.com/tokio-rs/tokio-uring/pull/100
[#107]: https://github.com/tokio-rs/tokio-uring/pull/107
[#111]: https://github.com/tokio-rs/tokio-uring/pull/111
[#113]: https://github.com/tokio-rs/tokio-uring/pull/113
[#116]: https://github.com/tokio-rs/tokio-uring/pull/116
[#121]: https://github.com/tokio-rs/tokio-uring/pull/121
[#124]: https://github.com/tokio-rs/tokio-uring/pull/124
[#126]: https://github.com/tokio-rs/tokio-uring/pull/126
[#130]: https://github.com/tokio-rs/tokio-uring/pull/130
[#131]: https://github.com/tokio-rs/tokio-uring/pull/131
[#132]: https://github.com/tokio-rs/tokio-uring/pull/132
[#133]: https://github.com/tokio-rs/tokio-uring/pull/133
[#134]: https://github.com/tokio-rs/tokio-uring/pull/134
[#135]: https://github.com/tokio-rs/tokio-uring/pull/135
[#136]: https://github.com/tokio-rs/tokio-uring/pull/136
[#139]: https://github.com/tokio-rs/tokio-uring/pull/139
[#141]: https://github.com/tokio-rs/tokio-uring/pull/141
[#142]: https://github.com/tokio-rs/tokio-uring/pull/142
[#143]: https://github.com/tokio-rs/tokio-uring/pull/143
[#144]: https://github.com/tokio-rs/tokio-uring/pull/144
[#146]: https://github.com/tokio-rs/tokio-uring/pull/146
[#148]: https://github.com/tokio-rs/tokio-uring/pull/148
[#149]: https://github.com/tokio-rs/tokio-uring/pull/149
[#153]: https://github.com/tokio-rs/tokio-uring/pull/153
[#154]: https://github.com/tokio-rs/tokio-uring/pull/154
[#155]: https://github.com/tokio-rs/tokio-uring/pull/155
[#157]: https://github.com/tokio-rs/tokio-uring/pull/157
[#162]: https://github.com/tokio-rs/tokio-uring/pull/162

### CODE REVIEW

**Comment by Noah-Kennedy** [comment]
@FrankReh I'd recommend making the description of this PR be your changelog updates. It makes it really easy to review and also to verify yourself if its rendered.

The main issue we hit with these things is bad links...

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,26 +1,192 @@
...
+- io: refactor post-op logic in ops into Completable ([#116])
+- by Noah-Kennedy
+
+- Support multi completion events: v2 ([#130])
+- by ollie-etl
+
+- simplify driver operation futures ([#139])
+- by FrankReh
+
+- rt: refactor runtime to avoid Rc&lt;RefCell&lt;...&gt;&gt; ([#142])
```
Comment (Noah-Kennedy) [inline CHANGELOG.md:37]: This got mangled

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,26 +1,192 @@
...
+
+- chore: add clippy to ci ([#100])
+- by Noah-Kennedy
+
+- ci: run cargo test --doc ([#135])
+- by ollie-etl
+
+### Unclassified so far
+
+- docs: all-features = true ([#84])
```
Comment (Noah-Kennedy) [inline CHANGELOG.md]: This is a fix

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,26 +1,192 @@
...
+- Add benchmarks for no-op submission ([#144])
+- by ollie-etl
+
+- Expose runtime structure ([#148])
+- by ollie-etl
+
+- Depend on io-uring version ^0.5.8 ([#153])
+- by FrankReh
+
+### Refactors
```
Comment (Noah-Kennedy) [inline CHANGELOG.md:32]: Change to `Internal Improvements`

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,26 +1,192 @@
+# 0.4.0 (November 5th, 2022)
+
+### Fixed
+
+- Fix panic in Deref/DerefMut for Slice extending into uninitialized part of the buffer ([#52])
+- by mzabaluev
+
+- chore: fix clippy lints ([#99])
```
Comment (Noah-Kennedy) [inline CHANGELOG.md:7]: This goes into `Internal Improvements`

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,26 +1,192 @@
...
+- Box the socket address to allow moving the Connect future ([#126])
+- by fasterthanlime
+
+- rt: Fix data race ([#146])
+- by ollie-etl
+
+
+### Added
+
+- Tokio-uring design proposal ([#1])
```
Comment (Noah-Kennedy) [inline CHANGELOG.md]: This can be omitted.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,26 +1,192 @@
...
+- by carllerche
+
+- driver: batch submit requests and add benchmark ([#78])
```
Comment (Noah-Kennedy) [inline CHANGELOG.md:14]: Add a `Changed` section and put this in there.

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,26 +1,192 @@
...
+- fs: implement FromRawFd for File ([#89])
+- by jiangliu
+
+- Implement `AsRawFd` for `TcpStream` ([#94])
+- by dpc
+
+- net: add TcpListener.local_addr method ([#107])
+- by FrankReh
+
+- write all ([#111])
```
Comment (Noah-Kennedy) [inline CHANGELOG.md:19]: This one got mangled on merge. Let's put this in a format that makes it understandable.

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,26 +1,192 @@
...
+- sync TcpStream, UnixStream and UdpSocket functionality ([#141])
+- by FrankReh
+
+- Add benchmarks for no-op submission ([#144])
+- by ollie-etl
+
+- Expose runtime structure ([#148])
+- by ollie-etl
+
+- Depend on io-uring version ^0.5.8 ([#153])
```
Comment (Noah-Kennedy) [inline CHANGELOG.md:30]: This goes under `Changed`

**Review by Noah-Kennedy — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (FrankReh) [reply CHANGELOG.md:37]: It actually displays okay in Markdown, according to the "display rich display" view. I was going to change it and then I thought maybe it should stay, as the '<' sign can cause markdown problems.

File: CHANGELOG.md (hunk 2 above)
Comment (FrankReh) [reply CHANGELOG.md]: I'm resolving as I make them locally.

File: CHANGELOG.md (hunk 7 above)
Comment (FrankReh) [reply CHANGELOG.md:19]: Made it net: add TcpStream.write_all

File: CHANGELOG.md (hunk 1 above)
Comment (Noah-Kennedy) [reply CHANGELOG.md:40]: I'd say that it should both display fine in markdown and also in text if possible.

Comment (FrankReh) [reply CHANGELOG.md:40]: Okay. I'll try with '<' and see what the rich display does with it. May require a backslash in front then.

**Comment by FrankReh** [comment]
I want to move #146 to be part of the Added #144, as it was a fix, but just for #144, not for anything seen in 0.3.0.

**Comment by Noah-Kennedy** [comment]
#146 was a fix for something discovered during work on #144 IIRC?

**Comment by Noah-Kennedy** [comment]
I thought it was an existing bug

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,26 +1,118 @@
...
+- net: add TcpStream.write_all ([#111])
+- driver: add Builder API as an option to start ([#113])
+- Socket and TcpStream shutdown ([#124])
+- fs: implement fs::File::from_std ([#131])
+- net: implement FromRawFd for TcpStream ([#132])
+- fs: implement OpenOptionsExt for OpenOptions ([#133])
+- Add NoOp support ([#134])
+- Add writev to TcpStream ([#136])
+- sync TcpStream, UnixStream and UdpSocket functionality ([#141])
+- Add benchmarks for no-op submission ([#144], [#146])
```
Comment (Noah-Kennedy) [inline CHANGELOG.md:25]: 146 fixed an existing issue that was uncovered by work on 144

**Comment by Noah-Kennedy** [comment]
Thanks!

**Review by Noah-Kennedy — APPROVED** [review]

**Pull request merged**